# Generate structural variants in a population
With a population generated by msprime (or any VCF), generate structural variants with VISOR and obtain a final VCF with structural variant sequences and population genotypes.

Help for commands is available:
```bash
python main.py -h
python main.py sv-vcf -h
python main.py msprime-vcf -h
```

## 1. Use a VCF as a base
A VCF is used as a base for genotypes and the position and number of variants. This VCF can be create with `msprime` to generate a population with:
```bash
python main.py msprime-vcf <FAI> <POPULATION SIZE> <MUTATION RATE> <SAMPLE SIZE> 
```
This command output as many VCF as there are chromosomes in the reference FAI.

## 2. Generate structural variants
Using the script `VISOR_random_bed/random_bed.sh`, generate a BED with structural variant informations. All structural variant types handled by VISOR are in the `VISOR_random_bed/visor_sv_type.yaml` file. Modify input in the script:
- generate as many variants as there are in the VCF
- choose structural variant types
- choose structural variant proportions (must equal 100)

[How to use the randomregion.r script](https://davidebolo1993.github.io/visordoc/usecases/usecases.html#visor-hack)

## 3. Get the final VCF
The final VCF is obtained by merging the msprime VCF, the VISOR BED and the structural variant sequences.
```bash
python main.py sv-vcf <BED> <VCF> <FASTA> <OUTNAME>s
```

Each variant type has a size distribution file (bins = 100 bp) in folder `sv_distributions`. The data was extracted from [An integrated map of structural variation in 2,504 human genomes (Sudmant, et al. 2015)](https://www.nature.com/articles/nature15394).

The distributions are used to randomly sample each structural variant size.

# TODO
- [ ] Add non supported VISOR variant types
    - [ ] MNP
    - [ ] tandem repeat contraction
    - [ ] tandem repeat expansion
    - [ ] approximate tandem repetition
- [ ] Add VCF merging when there are multiple chromosomes